### ISSUE TITLE
After logging in, the browser back button redirects users to a malformed URL with OnRedirecting component displayed forever

### ISSUE THREAD

**Issue by cfecherolle** [authts/react-oidc-context#1386]
Hi,

I've been having headaches trying to fix a bug I'm encountering when using `react-oidc-context` to authenticate with my Keycloak server.

> **Bug scenario**
> - Log in with username and password.
> - After being successfully redirected to the redirect URI, use the browser back button once.
> 
> **Expected result**
> - User is redirected to the redirect URI again (as they are already logged in)
> 
> **Actual result**
> - The user is redirected to a URL of the form :
> ```
> https://myapp.com/?state=88db7655520042d6a0b64e55895de911&session_state=99a1baa0-0637-4bc0-8c79-0ed229656e7a&code=adff040d-614e-476e-95a7-a08daded08ad.99a1baa0-0637-4bc0-8c79-0ed229656e7a.9e7b4e82-0244-42dd-8130-3dc2be1e2f23
> ```
> And the `OnRedirecting` component configured using `withAuthenticationRequired` is displayed forever.
> 

**Relevant info and configuration:**
My main component is configured like this:
```
export default withAuthenticationRequired(connectedRouter, {
    onBeforeSignin: () => {
        sessionStorage.setItem(AUTH_REDIRECT_URI_STORAGE_KEY, window.location.pathname);
    },
    OnRedirecting: () => <FullPageSpinner/>,
});
```
The whole React app is wrapped with:
```
<AuthProvider userManager={userManager} onSigninCallback={onSignInCallback}>
```
With props imported from:
```
import {UserManager, WebStorageStateStore} from "oidc-client-ts";

export const keycloakAuthority = `https://myapp.com/auth/realms/${import.meta.env.VITE_KEYCLOAK_REALM}`;
export const keycloakBackendClientId = "backend";

export const userManager = new UserManager({
    authority: keycloakAuthority,
    client_id: keycloakBackendClientId,
    monitorSession: true,
    redirect_uri: import.meta.env.VITE_KEYCLOAK_REDIRECT_URI,
    revokeTokensOnSignout: true,
    userStore: new WebStorageStateStore({store: window.sessionStorage}),
});

export const onSignInCallback = () => {
    const redirectPathname = sessionStorage.getItem(AUTH_REDIRECT_URI_STORAGE_KEY) || window.location.pathname;
    const redirectUrl = import.meta.env.VITE_KEYCLOAK_REDIRECT_URI + redirectPathname;
    window.location.replace(redirectUrl);
};
```
`AUTH_REDIRECT_URI_STORAGE_KEY` is the key of a session storage item I'm using to redirect users to their targeted URL after logging in.

The `window.location.replace` call in my sign in callback works in a way, as it redirects users to the redirect URI without any extraneous query params after logging in. Yet, when using the browser back button, this state is still present in history. I wish there was a way to either skip it (not adding it to history) or remove it after.

I would love to keep using this library as it was fast and easy to configure but if I can't get around this issue, this seems like a deal breaker for client-facing web apps. Any suggestions will be greatly appreciated 🙏

**Comment by cfecherolle** [comment]
I found this stale issue which seems to be related: https://github.com/authts/react-oidc-context/issues/683

**Comment by pamapa** [comment]
The `onSigninCallback` is thought to got next after sign-callback has been handled.  This seems you problem:

```ts
export const onSignInCallback = () => {
    const redirectPathname = sessionStorage.getItem(AUTH_REDIRECT_URI_STORAGE_KEY) || window.location.pathname;
    const redirectUrl = import.meta.env.VITE_KEYCLOAK_REDIRECT_URI + redirectPathname;
    window.location.replace(redirectUrl);
};
```
You put the redirectUrl into the location....

The solution should be like:
```ts
onSigninCallback: () => {
  // go to previous location
  const prevPath = this.getPreviousPath(); // to application url before the authetication redirect: e.g. http:// localhost/my-api/123
  window.location.replace(prevPath);
},
```

**Comment by cfecherolle** [comment]
Hi @pamapa, I'm not sure I understand what you mean.

When testing locally, my `import.meta.env.VITE_KEYCLOAK_REDIRECT_URI` is `http://localhost:3000` so what you suggest is exactly what I'm already doing: calling `window.location.replace` with `http://localhost:3000`+ whatever previous pathname the user was on, if any.

My problem is not with the `window.location.replace`, which works correctly and redirects my users like it should.
My problem lies with the previous history entry (the one with all the extraneous query params) the users end up with if they use the browser back button.

**Comment by pamapa** [comment]
> When testing locally, my `import.meta.env.VITE_KEYCLOAK_REDIRECT_URI` is `http://localhost:3000` so what you suggest is exactly what I'm already doing: calling `window.location.replace` with `http://localhost:3000`+ whatever previous pathname the user was on, if any.

Understood. Can you please try the following:
1. use `VITE_KEYCLOAK_REDIRECT_URI=http://localhost:3000/signin-callback` for the redirect. The signin-callback page must not exist, as this library intercepts via URI parameters.
2. use `window.location.replace("http://localhost:3000")` in `onSigninCallback`

That is how i use it and i have no back button issue. When i read https://developer.mozilla.org/en-US/docs/Web/API/Location/replace:
"The replace() method of the [Location](https://developer.mozilla.org/en-US/docs/Web/API/Location) interface replaces the current resource with the one at the provided URL. The difference from the [assign()](https://developer.mozilla.org/en-US/docs/Web/API/Location/assign) method is that after using replace() the current page will not be saved in session [History](https://developer.mozilla.org/en-US/docs/Web/API/History), meaning the **user won't be able to use the back button** to navigate to it."

**Comment by cfecherolle** [comment]
> > 
> Understood. Can you please try the following:
> 
> 1. use `VITE_KEYCLOAK_REDIRECT_URI=http://localhost:3000/signin-callback` for the redirect. The signin-callback page must not exist, as this library intercepts via URI parameters.
> 2. use `window.location.replace("http://localhost:3000")` in `onSigninCallback`
> 
> That is how i use it and i have no back button issue. When i read https://developer.mozilla.org/en-US/docs/Web/API/Location/replace: "The replace() method of the [Location](https://developer.mozilla.org/en-US/docs/Web/API/Location) interface replaces the current resource with the one at the provided URL. The difference from the [assign()](https://developer.mozilla.org/en-US/docs/Web/API/Location/assign) method is that after using replace() the current page will not be saved in session [History](https://developer.mozilla.org/en-US/docs/Web/API/History), meaning the **user won't be able to use the back button** to navigate to it."

I tried it but I was already using `window.location.replace` in `onSigninCallback` so it did not change the behavior in my case, even after changing the URL.
What browser(s) are you using, is your authentication server Keycloak, and if so, which version?

Thank you for taking the time to help with this!

**Comment by pamapa** [comment]
> What browser(s) are you using, is your authentication server Keycloak, and if so, which version?

Firefox and Chrome, both work as expected. My authentication server is Entra ID and not Keycloak.

**Comment by cfecherolle** [comment]
I think the problem I was having is related to the combined use of React Router with React OIDC Context. Manipulating the history manually might have been conflicting with the history handled by my BrowserRouter.

To work with this, I needed the Router to be aware of the Auth (to secure acess) and the Auth to be aware of the Router (to 
manipulate the history in `onSignInCallback`), which was impossible, as one necessarily needs to wrap the other.

My App is structured as follows (simplified):
```
    <AuthProvider userManager={userManager}>
        <StrictMode>
            <BrowserRouter>
                <Switch>
                    <Route path="/" render={(props) => <MainRouter {...props}}/>
                </Switch>
            </BrowserRouter>
        </StrictMode>
    </AuthProvider>
```

I found a workaround to this by moving the sign-in callback logic to its own component.
Here is what I did:

1. Change `redirect_uri` to `${import.meta.env.VITE_KEYCLOAK_REDIRECT_URI}/signin-callback`
2. Remove `onSigninCallback` prop from `AuthProvider`
3. Add a new `Route` to my `Switch`: `<Route path="/signin-callback" component={SignInCallback}/>`
4. Create a new `SignInCallback` component to handle redirection:
```
const SignInCallbackComponent: React.FC<RouteComponentProps> = ({history}) => {
    const { isLoading} = useAuth();
    useEffect(() => {
        if (hasAuthParams() && !isLoading) {
            const redirectPathname = sessionStorage.getItem(AUTH_REDIRECT_URI_STORAGE_KEY) || "/";
            history.replace(redirectPathname);
        }
    }, [isLoading, history]);
    return <FullPageSpinner/>;
};

export const SignInCallback = withRouter(SignInCallbackComponent);
```
This new component is aware of both Auth and Router and can handle the redirections correctly.

The only slight issue left is that when I use the browser back button after login, I see the URL going back and forth twice (query params -> no query params -> query params -> no query params). This is acceptable, as the user ends up on the right page and not a dead-end.

Thank you for your help and suggestions!
